plugin/federated/CMakeLists.txt

```diff
@@ -1,26 +1,48 @@
 # gRPC needs to be installed first. See README.md.
+set(protobuf_MODULE_COMPATIBLE TRUE)
+set(protobuf_BUILD_SHARED_LIBS TRUE)
 find_package(Protobuf CONFIG REQUIRED)
 find_package(gRPC CONFIG REQUIRED)
-find_package(Threads)
+message(STATUS "Found gRPC: ${gRPC_CONFIG}")
+set(_PROTOBUF_LIBPROTOBUF protobuf::libprotobuf)
+set(_PROTOBUF_PROTOC $<TARGET_FILE:protobuf::protoc>)
+set(_GRPC_GRPCPP gRPC::grpc++)
+set(_GRPC_CPP_PLUGIN_EXECUTABLE $<TARGET_FILE:gRPC::grpc_cpp_plugin>)
+
+# Proto file
+get_filename_component(federated_proto "federated.proto" ABSOLUTE)
+get_filename_component(federated_proto_path "${federated_proto}" PATH)
 
 # Generated code from the protobuf definition.
-add_library(federated_proto federated.proto)
-target_link_libraries(federated_proto PUBLIC protobuf::libprotobuf gRPC::grpc gRPC::grpc++)
-target_include_directories(federated_proto PUBLIC ${CMAKE_CURRENT_BINARY_DIR})
-set_property(TARGET federated_proto PROPERTY POSITION_INDEPENDENT_CODE ON)
+set(federated_srcs "${CMAKE_CURRENT_BINARY_DIR}/federated.pb.cc")
+set(federated_hdrs "${CMAKE_CURRENT_BINARY_DIR}/federated.pb.h")
+set(federated_grpc_srcs "${CMAKE_CURRENT_BINARY_DIR}/federated.grpc.pb.cc")
+set(federated_grpc_hdrs "${CMAKE_CURRENT_BINARY_DIR}/federated.grpc.pb.h")
+add_custom_command(
+      OUTPUT "${federated_srcs}" "${federated_hdrs}"
+             "${federated_grpc_srcs}" "${federated_grpc_hdrs}"
+      COMMAND ${_PROTOBUF_PROTOC}
+      ARGS --grpc_out "${CMAKE_CURRENT_BINARY_DIR}"
+           --cpp_out "${CMAKE_CURRENT_BINARY_DIR}"
+           -I "${federated_proto_path}"
+           --plugin=protoc-gen-grpc="${_GRPC_CPP_PLUGIN_EXECUTABLE}"
+           "${federated_proto}"
+      DEPENDS "${federated_proto}")
 
-get_target_property(grpc_cpp_plugin_location gRPC::grpc_cpp_plugin LOCATION)
-protobuf_generate(TARGET federated_proto LANGUAGE cpp)
-protobuf_generate(
-    TARGET federated_proto
-    LANGUAGE grpc
-    GENERATE_EXTENSIONS .grpc.pb.h .grpc.pb.cc
-    PLUGIN "protoc-gen-grpc=${grpc_cpp_plugin_location}")
+add_library(federated_proto_lib STATIC)
+target_sources(federated_proto_lib PRIVATE
+  ${federated_srcs} ${federated_hdrs}
+  ${federated_grpc_srcs} ${federated_grpc_hdrs})
+target_link_libraries(federated_proto_lib PUBLIC
+  ${_PROTOBUF_LIBPROTOBUF} ${_GRPC_GRPCPP})
+target_include_directories(federated_proto_lib PUBLIC
+  ${CMAKE_CURRENT_BINARY_DIR})
+xgboost_target_properties(federated_proto_lib)
 
```

Comment:
Are these changes still necessary now that compiling grpc is no longer part of CMake?

Reply:
I reverted the changes.